engine/table/src/main/java/io/deephaven/engine/table/impl/sources/ring/AbstractRingChunkSource.java

```diff
@@ -0,0 +1,389 @@
+                physicalRows.close();
+            }
+        }
+        nextRingIx += logicalFillSize;
+    }
+
+    @Override
+    public final Chunk<Values> getChunk(@NotNull GetContext context, @NotNull RowSequence rowSequence) {
+        if (rowSequence.isEmpty()) {
+            return getChunkType().getEmptyChunk();
+        }
+        if (rowSequence.isContiguous()) {
+            return getChunk(context, rowSequence.firstRowKey(), rowSequence.lastRowKey());
+        }
+        final WritableChunk<Values> chunk = DefaultGetContext.getWritableChunk(context);
+        fillChunk(DefaultGetContext.getFillContext(context), chunk, rowSequence);
+        return chunk;
+    }
+
+    @Override
+    public final Chunk<Values> getChunk(@NotNull GetContext context, long firstKey, long lastKey) {
+        // This check should not be necessary given precondition on getChunk
+        if (!containsRange(firstKey, lastKey)) {
+            throw new IllegalStateException(
+                    String.format("getChunk precondition broken, invalid range. requested=[%d, %d], available=[%d, %d]",
+                            firstKey, lastKey, firstKey(), lastKey()));
+        }
+        final int firstRingIx = keyToRingIndex(firstKey);
+        final int secondRingIx = keyToRingIndex(lastKey);
+        if (firstRingIx <= secondRingIx) {
+            // Optimization when we can return a contiguous view
+            return ring(firstRingIx, secondRingIx - firstRingIx + 1); // todo: is it bad that this is writable?
+        }
+        final WritableChunk<Values> chunk = DefaultGetContext.getWritableChunk(context);
+        final int size = fillByCopy2(chunk, 0, firstRingIx, secondRingIx);
+        if (size != lastKey - firstKey + 1) {
+            throw new IllegalStateException();
+        }
+        chunk.setSize(size);
+        return chunk;
+    }
+
+    @Override
+    public final void fillChunk(@NotNull FillContext context, @NotNull WritableChunk<? super Values> destination,
+            @NotNull RowSequence rowSequence) {
+        if (rowSequence.getAverageRunLengthEstimate() < USE_RANGES_AVERAGE_RUN_LENGTH) {
+            try (final KeyFiller filler = new KeyFiller(destination)) {
+                rowSequence.forAllRowKeys(filler);
+            }
+        } else {
+            try (final RangeFiller filler = new RangeFiller(destination)) {
+                rowSequence.forAllRowKeyRanges(filler);
+            }
+        }
+    }
+
+    private final class KeyFiller implements LongConsumer, Closeable {
+        private final WritableChunk<? super Values> destination;
+        private int destOffset;
+
+        public KeyFiller(WritableChunk<? super Values> destination) {
+            this.destination = Objects.requireNonNull(destination);
+        }
+
+        @Override
+        public void accept(long key) {
+            // this check should not be necessary given precondition on fillChunk
+            if (!containsKey(key)) {
```

Comment:
I would not spend the cycles on this check. Better to delete it, or put it under a debug flag. Same with `containsRange` in `RangeFiller`.

Reply:
I'm able to move this check up to the fillChunk method.